### PR TITLE
[Clang] Re-land Overflow Pattern Exclusions

### PR DESCRIPTION
**Pull request by JustinStitt**
[[Clang] Overflow Pattern Exclusions #100272
](https://github.com/llvm/llvm-project/pull/100272) was merged last week but had some test failures caused by an uninitialized bit within a bitfield. These failures have been fixed so let's try to re-land this.

After speaking with Vitaly and Kees, we've decided to switch the flag naming to better match the existing conventions of sanitizer options: `-fsanitize-<sanitizer_kind>-<option>`

old flag:
`-fsanitize-overflow-pattern-exclusion=(all|none|...)`

new flag:
`-fsanitize-undefined-ignore-overflow-pattern=(all|none|...)`

cc: @vitalybuka @kees @efriedma-quic @bwendling

### CODE REVIEW

**Comment by JustinStitt** [comment]
@bwendling, obviously this all has to be squashed into one commit when merged -- feel free to use the previous commit log from https://github.com/llvm/llvm-project/commit/9a666deecb9ff6ca3a6b12e6c2877e19b74b54da but change the flag to match the new one 👍

File: clang/lib/Driver/SanitizerArgs.cpp [hunk 1]
```diff
@@ -119,6 +119,10 @@ static SanitizerMask parseArgValues(const Driver &D, const llvm::opt::Arg *A,
 static int parseCoverageFeatures(const Driver &D, const llvm::opt::Arg *A,
                                  bool DiagnoseErrors);
 
+static int parseOverflowPatternExclusionValues(const Driver &D,
```
[the remark this answers is not in the document]
Comment (JustinStitt) [reply clang/lib/Driver/SanitizerArgs.cpp:124]: Yep. [03f5d81](https://github.com/llvm/llvm-project/pull/104889/commits/03f5d81374784377d382eef7286d74eb712c7395)

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -438,6 +438,36 @@ Moved checkers
 Sanitizers
 ----------
 
+- Added the ``-fsanitize-undefined-ignore-overflow-pattern`` flag which can be
+  used to disable specific overflow-dependent code patterns. The supported
+  patterns are: ``add-overflow-test``, ``negated-unsigned-const``, and
+  ``post-decr-while``. The sanitizer instrumentation can be toggled off for all
```
Comment (vitalybuka) [inline clang/docs/ReleaseNotes.rst:444]: it does not state what is the default state of the flag.

File: clang/docs/UndefinedBehaviorSanitizer.rst [hunk 3]
```diff
@@ -293,6 +293,48 @@ To silence reports from unsigned integer overflow, you can set
...
+    /// -fsanitize-undefined-ignore-overflow-pattern=negated-unsigned-const
+    unsigned long foo = -1UL; // No longer causes a negation overflow warning
+    unsigned long bar = -2UL; // and so on...
+
+``post-decr-while``
+
+.. code-block:: c++
+
+    /// -fsanitize-undefined-ignore-overflow-pattern=post-decr-while
+    unsigned char count = 16;
```
Comment (vitalybuka) [inline clang/docs/UndefinedBehaviorSanitizer.rst:322]: does this applies to unsigned only or signed as well?

File: clang/docs/UndefinedBehaviorSanitizer.rst [hunk 4]
```diff
@@ -293,6 +293,48 @@ To silence reports from unsigned integer overflow, you can set
...
+    while (count--) { /* ... */ } // No longer causes unsigned-integer-overflow sanitizer to trip
+
+``add-overflow-test``
+
+.. code-block:: c++
+
+    /// -fsanitize-undefined-ignore-overflow-pattern=add-overflow-test
+    if (base + offset < base) { /* ... */ } // The pattern of `a + b < a`, and other re-orderings,
```
Comment (vitalybuka) [inline clang/docs/UndefinedBehaviorSanitizer.rst:330]: why is that needed, looks like a terrible UB?
CC @eugenis

File: clang/docs/UndefinedBehaviorSanitizer.rst [hunk 5]
```diff
@@ -293,6 +293,48 @@ To silence reports from unsigned integer overflow, you can set
...
+
+``post-decr-while``
+
+.. code-block:: c++
+
+    /// -fsanitize-undefined-ignore-overflow-pattern=post-decr-while
+    unsigned char count = 16;
+    while (count--) { /* ... */ } // No longer causes unsigned-integer-overflow sanitizer to trip
+
+``add-overflow-test``
```
Comment (vitalybuka) [inline clang/docs/UndefinedBehaviorSanitizer.rst:325]: with `fsanitize-undefined-ignore-overflow-pattern=` name
another `overflow` seems redundant in name

File: clang/docs/ReleaseNotes.rst (hunk 2 above)
Comment (JustinStitt) [reply clang/docs/ReleaseNotes.rst:444]: It is not explicitly stated but no patterns are excluded by default. This is somewhat evident by "... _can be_ used to disable" and "sanitizer instrumentation _can be_ toggled off".

File: clang/include/clang/Basic/LangOptions.h [hunk 6]
```diff
@@ -367,6 +367,21 @@ class LangOptionsBase {
...
 
+  /// Exclude certain code patterns from being instrumented by arithmetic
+  /// overflow sanitizers
+  enum OverflowPatternExclusionKind {
+    /// Don't exclude any overflow patterns from sanitizers
+    None = 1 << 0,
+    /// Exclude all overflow patterns (below)
+    All = 1 << 1,
+    /// if (a + b < a)
+    AddOverflowTest = 1 << 2,
```
Comment (vitalybuka) [inline clang/include/clang/Basic/LangOptions.h:378]: `AddTest`

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 5 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:325]: The pattern itself is an "overflow checking" pattern that just so happens to trigger an overflow.

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 4 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:330]: Why is what needed? Maybe the GitHub UI is failing me but I can't exactly tell what you're referring to.

File: clang/docs/UndefinedBehaviorSanitizer.rst [hunk 7]
```diff
@@ -293,6 +293,48 @@ To silence reports from unsigned integer overflow, you can set
...
+
+There are certain overflow-dependent or overflow-prone code patterns which
+produce a lot of noise for integer overflow/truncation sanitizers. Negated
+unsigned constants, post-decrements in a while loop condition and simple
+overflow checks are accepted and pervasive code patterns. However, the signal
+received from sanitizers instrumenting these code patterns may be too noisy for
+some projects. To disable instrumentation for these common patterns one should
+use ``-fsanitize-undefined-ignore-overflow-pattern=``.
+
+Currently, this option supports three overflow-dependent code idioms:
```
Comment (vitalybuka) [inline clang/docs/UndefinedBehaviorSanitizer.rst:307]: Can you please add a table to show which which patter applies to which ubsan check?

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 3 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:322]: just unsigned, the comment below tries to describe this behavior:

`// No longer causes unsigned-integer-overflow sanitizer to trip`

File: clang/test/CodeGen/ignore-overflow-pattern-false-pos.c [hunk 8]
```diff
@@ -0,0 +1,63 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fsanitize=signed-integer-overflow,unsigned-integer-overflow -fsanitize-undefined-ignore-overflow-pattern=all %s -emit-llvm -o - | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fsanitize=signed-integer-overflow,unsigned-integer-overflow -fsanitize-undefined-ignore-overflow-pattern=all -fwrapv %s -emit-llvm -o - | FileCheck %s
+
+// Check for potential false positives from patterns that _almost_ match classic overflow-dependent or overflow-prone code patterns
+extern unsigned a, b, c;
+extern int u, v, w;
```
Comment (vitalybuka) [inline clang/test/CodeGen/ignore-overflow-pattern-false-pos.c:6]: are these unused?

File: clang/docs/ReleaseNotes.rst (hunk 2 above)
Comment (vitalybuka) [reply clang/docs/ReleaseNotes.rst:444]: I understand. I am actually asking to add explicit statement. E.g. `The dafault is none`.
or none (default)

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 3 above)
Comment (vitalybuka) [reply clang/docs/UndefinedBehaviorSanitizer.rst:322]: than naming is inconsistent, one above have unsigned in name

Comment (vitalybuka) [reply clang/docs/UndefinedBehaviorSanitizer.rst:322]: One above has unsigned in name, this does not. Let's rename for consistency?

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 5 above)
Comment (vitalybuka) [reply clang/docs/UndefinedBehaviorSanitizer.rst:325]: That's too deep :)

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 4 above)
Comment (vitalybuka) [reply clang/docs/UndefinedBehaviorSanitizer.rst:330]: Sorry, I assumed it attached to the check.

Suppressing overflow checks on `base + offset < base`  with signed types.

File: clang/include/clang/Basic/LangOptions.h (hunk 6 above)
Comment (vitalybuka) [reply clang/include/clang/Basic/LangOptions.h:378]: As above

File: clang/docs/ReleaseNotes.rst (hunk 2 above)
Comment (JustinStitt) [reply clang/docs/ReleaseNotes.rst:444]: Ok. I will open a PR to expand the documentation following your review.

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 3 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:322]: `unsigned-post-decr-while`?

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 4 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:330]: Since the overflow checks are never inserted with `add-overflow-test` in that case the optimizer would be free to do whatever it wants. One should consider `-fwrapv`. These pattern exclusions should not try to do too much (i.e; try to prevent the optimizer from doing its thing).

File: clang/docs/UndefinedBehaviorSanitizer.rst (hunk 7 above)
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:307]: Yes, good idea.

File: clang/test/CodeGen/ignore-overflow-pattern-false-pos.c (hunk 8 above)
Comment (JustinStitt) [reply clang/test/CodeGen/ignore-overflow-pattern-false-pos.c:6]: u, v, and we are unused. Removing them in a follow-up PR (also touching-up documentation). 

Thanks
